Publish Multi-Arch images using github actions

wickywaka opened the pull request:
I manage to publish multi-arch images using GitHub Actions and here is the merge request. There are still open questions and I was doing it while learning at the same time so please suggest what can be improved.

It uses GitHub secrets for docker hub username and token with following names DOCKER_HUB_USERNAME, DOCKER_HUB_TOKEN. Please the docker.yml file for more details.

Related Issue: #23

robinsmidsrod commented:
@waqarrashid33 This is a great start! Now comes the comments. ;)

1. Any reason why the build environment uses ubuntu 18.04 instead of the newer 20.04 or 22.04? Are the steps not compatible with the newest OS versions?
2. I noticed you're just directly pushing the `:latest` tag. Is there any way to push a specific version, or push a build-artifact named after the digest? I'd like to be able to push dev versions of the images so it is easier to collaborate on testing before doing a new release.
3. I found the secrets setup in **Settings**, so that part should be okay, but when does this workflow trigger? I'm assuming it triggers when code enters the master branch. How would this work for numbered releases. How about recurring rebuilds when the base ubuntu image updates?
4. Just curious: Do you need QEMU when making an x86/x64 image? Is this just the best way to create multiarch images?

As you can probably notice, I'm not familiar with GitHub Actions, hence all the questions. I have worked some with Bitbucket Pipelines, so CI/CD is not completely unfamiliar to me.

wickywaka commented:
For just running the existing image, you can find it on docker hub
wickywaka/docker-dhcpd.

On Wed, Jun 1, 2022, 20:31 Waqar Rashid ***@***.***> wrote:

> Hi Robin,
> I haven't given it much thought consider this as an MVP.
>
> 1. No reason, I think 20.04 should be fine. 22.04 is too new for my taste.
>
> 2. From what I have read, we can change the trigger when make/push a tag
> in GitHub. It can also be so that it takes the tag as tag for docker image
> e.g docker-dhcpd:3.2 etc. There are lots of possibilities in this area.
> This is my first time using GitHub Actions.
>
> 3. Good question, I don't know how to update when base image changes.
>
> 4. I think it uses qemu only for building images whose architecture is
> different than host. On my system the amd64 image was always finishing its
> build first because it was probably running natively.
>
>
> On Wed, Jun 1, 2022, 16:46 Robin Smidsrød ***@***.***>
> wrote:
>
>> @waqarrashid33 <https://github.com/waqarrashid33> This is a great start!
>> Now comes the comments. ;)
>>
>>    1. Any reason why the build environment uses ubuntu 18.04 instead of
>>    the newer 20.04 or 22.04? Are the steps not compatible with the newest OS
>>    versions?
>>    2. I noticed you're just directly pushing the :latest tag. Is there
>>    any way to push a specific version, or push a build-artifact named after
>>    the digest? I'd like to be able to push dev versions of the images so it is
>>    easier to collaborate on testing before doing a new release.
>>    3. I found the secrets setup in *Settings*, so that part should be
>>    okay, but when does this workflow trigger? I'm assuming it triggers when
>>    code enters the master branch. How would this work for numbered releases.
>>    How about recurring rebuilds when the base ubuntu image updates?
>>    4. Just curious: Do you need QEMU when making an x86/x64 image? Is
>>    this just the best way to create multiarch images?
>>
>> As you can probably notice, I'm not familiar with GitHub Actions, hence
>> all the questions. I have worked some with Bitbucket Pipelines, so CI/CD is
>> not completely unfamiliar to me.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/networkboot/docker-dhcpd/pull/31#issuecomment-1143705560>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABVDYA5LOKQI2M6JL6ZYPJ3VM5ZT7ANCNFSM5XQMTWBA>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

wickywaka commented:
Hi @robinsmidsrod ,
I haven't given it much thought consider this as an MVP. Here I will post my answers to your questions/comments again (somehow email reply got lost).

> 1. Any reason why the build environment uses ubuntu 18.04 instead of the newer 20.04 or 22.04? Are the steps not compatible with the newest OS versions?

No reason, I think 20.04 should be fine. 22.04 is too new for my taste.

> 2. I noticed you're just directly pushing the `:latest` tag. Is there any way to push a specific version, or push a build-artifact named after the digest? I'd like to be able to push dev versions of the images so it is easier to collaborate on testing before doing a new release.

From what I have read about GitHub Actions, we can change the trigger when make/push a tag in GitHub. It can also be so that it takes the tag as tag for docker image e.g docker-dhcpd:3.2 etc. There are lots of possibilities in this area. This is my first time using GitHub Actions. So definitely some work to do here.

> 3. I found the secrets setup in **Settings**, so that part should be okay, but when does this workflow trigger? I'm assuming it triggers when code enters the master branch. How would this work for numbered releases. How about recurring rebuilds when the base ubuntu image updates?

 3. Good question, I don't know how to update when base image changes or what would be a good strategy. 

> 4. Just curious: Do you need QEMU when making an x86/x64 image? Is this just the best way to create multiarch images?

I think it uses qemu only for building images whose architecture is different than host. On my system the amd64 image was always finishing its build first because it was probably running natively.
 
> As you can probably notice, I'm not familiar with GitHub Actions, hence all the questions. I have worked some with Bitbucket Pipelines, so CI/CD is not completely unfamiliar to me.

I am new to this as well but we can learn from each other.

katbyte commented:
found this while trying to stand up docker dns & dhcp on a rpi, look forward to this landing on docker hub!

what you could do is have 1 job that runs on every commit and pushes to latest, and then another that runs on version tags:
```
on:
  push:
    tags:
      - 'v[0-9]+.[0-9]+.[0-9]+*'
```

and then grab the tag with `${{ steps.vars.outputs.tag }}` or `${{ env.RELEASE_VERSION }}` and pop that into the docker tag name (and even do multiple base versions of the os like here https://hub.docker.com/r/ubuntu/bind9/tags)

yscialom commented:
Hello, and thank your for this PR.

Tagging the result of `master`  as `:latest` is a good idea. Additionally, tagging the version number would be a perfect solution.

This would require to run the workflow on git tags (1) and to extract the git's tag name to construct the docker's tag name (2).

---

Example for (1)
```yaml
on:
  push:
    tags:
      - *
```
See [Github's workflow syntax documentation](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-branches-and-tags).

---

Example for (2)
```yaml
    steps:
      - uses: olegtarasov/get-tag@v2.1
        id: tagName
        with:
          tagRegex: "foobar-(.*)"  # Optional. Returns specified group text as tag name. Full tag string is returned if regex is not defined.
          tagRegexGroup: 1 # Optional. Default is 1.
      - name: Some other step # Output usage example
        with:
          tagname: ${{ steps.tagName.outputs.tag }}
      - name: Yet another step # Environment variabl usage example
        run: |
          docker build . --file Dockerfile --tag docker.pkg.github.com/someimage:$GIT_TAG_NAME
```
See [actions/get-tag-name documentation](https://github.com/marketplace/actions/get-tag-name).

Additionally<sup>2</sup>, it would be nice if the publishing of git tag x.y.z would publish the docker tags `:x` `:x.y` _and_ `:x.y.z`.

wickywaka commented:
Hey guys,
Thanks for the suggestions. I will try to add them to the PR as soon as I can.

yscialom commented:
@wickywaka 
Hello. I'd be happy to see this PR merged in the near future. How can I help?

wickywaka commented:
> @wickywaka Hello. I'd be happy to see this PR merged in the near future. How can I help?

@yscialom Lets see if I can finish it up today. Anyway I have added you as collaborator to the fork so that you can make the desired changes into the merge request.

wickywaka commented:
@robinsmidsrod @katbyte @yscialom
Hey guys,
Here are some changes:

1. Updated the image that is used to build these docker images to ubuntu-latest.
2. Push image on tag labelled as tag (Need Feedback)
3. Changed names (Need feedback)

So now we are pushing :latest on every commit to master and pushing :vX.X.X on tag. Have a look at these files and recommend changes.

Waqar

robinsmidsrod commented on .github/workflows/dockerize-latest.yml:
```diff
@@ -0,0 +1,40 @@
+  workflow_dispatch:
+
+jobs:
+  build-Push-Latest:
```
Consistent casing. I prefer just to use small letters with dashes here.

robinsmidsrod commented on .github/workflows/dockerize-latest.yml:
```diff
@@ -0,0 +1,40 @@
+        with:
+          platforms: linux/amd64,linux/arm/v7,linux/arm64
+          push: true
+          tags: wickywaka/docker-dhcpd:latest
```
This must be changed to the correct Docker hub repo before final merge.

robinsmidsrod commented on .github/workflows/dockerize-tag.yml:
```diff
@@ -0,0 +1,41 @@
+        with:
+          platforms: linux/amd64,linux/arm/v7,linux/arm64
+          push: true
+          tags: wickywaka/docker-dhcpd:${{ steps.tagName.outputs.tag }}
```
This must be changed to the correct Docker hub repo before final merge.

robinsmidsrod commented on .github/workflows/dockerize-tag.yml:
```diff
@@ -0,0 +1,41 @@
+    runs-on: ubuntu-latest
+    steps:
+      - uses: olegtarasov/get-tag@v2.1
+        id: tagName
```
Use consistent casing. Since all the other ids are snake case, this one should be too. Recommend `get_tag` instead of tagName.

robinsmidsrod replied:
Also, I'd prefer if the Docker tag doesn't contain the letter `v` at the start. So add a tagRegex to strip that out, as shown on the doc page for this step.

wickywaka replied:
> Also, I'd prefer if the Docker tag doesn't contain the letter `v` at the start. So add a tagRegex to strip that out, as shown on the doc page for this step.

I am not so good at regex so I might need a bit more help.  Do I add the following:

```
 tagRegex: "v-(.*)"
```

robinsmidsrod replied:
I would assume that `^v(\d+\.\d+\.\d+)$` is what you're looking for, which will match a semver-style version number.

wickywaka replied:
> I would assume that `^v(\d+\.\d+\.\d+)$` is what you're looking for, which will match a semver-style version number.

Ok. Before that I will try to use https://github.com/marketplace/actions/parse-semver, I think this will make the future tagging much easier. For example, tags like  :1, :1.1, :1.1.1 etc.

Currently, with parse-semver action, it fails to run when I push a tag, may be due to some syntax error in yml or something else.

robinsmidsrod replied:
I think you'll need need to put the `get-tag` step first, and use the `$GIT_TAG_NAME` variable you get from that as the sem-ver input string. Because right now it looks like you're trying to extract the semver fields out of a static string, not the actual tag you just tagged with your git client.

wickywaka replied:
I think it should be ready now.

robinsmidsrod reviewed (changes requested):
This is close to ready for primetime. Just a few more changes and it should be usable.

robinsmidsrod commented:
@wickywaka Great work! This is coming along nicely! Good job on getting the semver version parsed correctly. I do notice two things that should probably be fixed before we merge it.

1. Can you add a step to build and push the Dockerfile.ldap to both files as well? It would be unfortunate if that is not built with automation when the rest is. Tag it as `:ldap-latest` and `:ldap-X.Y.Z`.
2. Since you were able to add semver parsing, whould it be possible to also build tags `:1` (major semver version number) and `:ldap-1`? This is for people that want to use the this tag in their scenarios and avoid possibly breaking changes (that we would know about).

wickywaka commented:
> @wickywaka Great work! This is coming along nicely! Good job on getting the semver version parsed correctly. I do notice two things that should probably be fixed before we merge it.
> 
> 1. Can you add a step to build and push the Dockerfile.ldap to both files as well? It would be unfortunate if that is not built with automation when the rest is. Tag it as `:ldap-latest` and `:ldap-X.Y.Z`.
> 2. Since you were able to add semver parsing, whould it be possible to also build tags `:1` (major semver version number) and `:ldap-1`? This is for people that want to use the this tag in their scenarios and avoid possibly breaking changes (that we would know about).

1. I don't know much about the ldap stuff. I don't even know what it is used for, so I can use some pointers here.
2. Tagging won't be a problem. The question I have here: Do we tag :1 with the latest one in this major release. For example: If :1 points to :1.1.1 but a new tag was pushed with 1.2.1, then we would have to update the tag :1 to :1.2.1. If the behavior you want is a simple one like if we push a tag 1.1.2 then we make three tags :1, :1.1 and :1.1.2, overwriting the previous :1 and :1.1. Is it desired behavior? I hope I was able to explain myself :smile:

robinsmidsrod commented:
1. The LDAP stuff is just another Dockerfile, so you just need to figure out how to specify something else than the default `Dockerfile` to the build-push step. I'm going to assume that is an option in there.
2. I thought about pushing both `:X`, `:X.Y` and `:X.Y.Z` tags, but from a semver perspective it is not needed, because only major versions should introduce breaking changes. So we only need `:latest`, `:X`, and `:X.Y.Z` and similar with the `ldap-` prefix. And yes, when we tag a new version that should overwrite the `:X` tag on Dockerhub pointing to the last released of that major version.

wickywaka commented:
> 1. The LDAP stuff is just another Dockerfile, so you just need to figure out how to specify something else than the default `Dockerfile` to the build-push step. I'm going to assume that is an option in there.


Can you check if this is what you expected:
` docker pull wickywaka/docker-dhcpd:ldap-latest`

wickywaka commented:
> 2. I thought about pushing both `:X`, `:X.Y` and `:X.Y.Z` tags, but from a semver perspective it is not needed, because only major versions should introduce breaking changes. So we only need `:latest`, `:X`, and `:X.Y.Z` and similar with the `ldap-` prefix. And yes, when we tag a new version that should overwrite the `:X` tag on Dockerhub pointing to the last released of that major version.

I think its ready as well.

robinsmidsrod commented on .github/workflows/dockerize-tag.yml:
```diff
@@ -0,0 +1,66 @@
+          push: true
+          file: Dockerfile.ldap
+          tags:
+            networkboot/dhcpd:ladp-${{ steps.semver-parser.outputs.fullversion }},
```
Typo. `ladp` -> `ldap`.

wickywaka replied:
Good catch. It should be fixed now.

yscialom commented:
Congrats!
![image](https://user-images.githubusercontent.com/1231051/183696941-ca36384f-e220-4c82-b5fe-15b81b1873de.png)

wickywaka commented:
Thanks. I will try to add this.

On Sun, Jun 5, 2022, 20:32 kt ***@***.***> wrote:

> found this while trying to stand up docker dns & dhcp on a rpi, look
> forward to this landing on docker hub!
>
> what you could do is have 1 job that runs on every commit and pushes to
> latest, and then another that runs on version tags:
>
> on:
>   push:
>     tags:
>       - 'v[0-9]+.[0-9]+.[0-9]+*'
>
> and then grab the tag with ${{ steps.vars.outputs.tag }} or ${{
> env.RELEASE_VERSION }} and pop that into the docker tag name (and even do
> multiple base versions of the os like here
> https://hub.docker.com/r/ubuntu/bind9/tags)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/networkboot/docker-dhcpd/pull/31#issuecomment-1146862050>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVDYA6ALT4UTXZK5ZFT72LVNTXC7ANCNFSM5XQMTWBA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

wickywaka commented:
Hi Robin,
I haven't given it much thought consider this as an MVP.

1. No reason, I think 20.04 should be fine. 22.04 is too new for my taste.

2. From what I have read, we can change the trigger when make/push a tag in
GitHub. It can also be so that it takes the tag as tag for docker image e.g
docker-dhcpd:3.2 etc. There are lots of possibilities in this area. This is
my first time using GitHub Actions.

3. Good question, I don't know how to update when base image changes.

4. I think it uses qemu only for building images whose architecture is
different than host. On my system the amd64 image was always finishing its
build first because it was probably running natively.


On Wed, Jun 1, 2022, 16:46 Robin Smidsrød ***@***.***> wrote:

> @waqarrashid33 <https://github.com/waqarrashid33> This is a great start!
> Now comes the comments. ;)
>
>    1. Any reason why the build environment uses ubuntu 18.04 instead of
>    the newer 20.04 or 22.04? Are the steps not compatible with the newest OS
>    versions?
>    2. I noticed you're just directly pushing the :latest tag. Is there
>    any way to push a specific version, or push a build-artifact named after
>    the digest? I'd like to be able to push dev versions of the images so it is
>    easier to collaborate on testing before doing a new release.
>    3. I found the secrets setup in *Settings*, so that part should be
>    okay, but when does this workflow trigger? I'm assuming it triggers when
>    code enters the master branch. How would this work for numbered releases.
>    How about recurring rebuilds when the base ubuntu image updates?
>    4. Just curious: Do you need QEMU when making an x86/x64 image? Is
>    this just the best way to create multiarch images?
>
> As you can probably notice, I'm not familiar with GitHub Actions, hence
> all the questions. I have worked some with Bitbucket Pipelines, so CI/CD is
> not completely unfamiliar to me.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/networkboot/docker-dhcpd/pull/31#issuecomment-1143705560>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVDYA5LOKQI2M6JL6ZYPJ3VM5ZT7ANCNFSM5XQMTWBA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
